### PR TITLE
docs: document the compilation unit digesting algorithm

### PR DESCRIPTION
**Pull request by jaysachs**
This is just an attempt to document what we currently have implemented. I can imagine simpler, cleaner algorithms but it's functional.

Addresses #3680

### CODE REVIEW

File: kythe/docs/kythe-kzip.txt [hunk 1]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+=== Computing the digest for a Compilation Unit
+
+The representative implementation of this is in
+`kythe/go/platform/kcd/kythe/units.go`.
+
+Definitions:
+   `NULL`: the one-byte value corresponding to ASCII NULL (`\0x00`)
+   `NL`:   the one-byte value corresponding to ASCII newline (`\x0a`)
+
+Any strings are emitted as the as described by a UTF-8 encoding.
```
Comment (justbuchanan) [inline kythe/docs/kythe-kzip.txt]: s/as the//

File: kythe/docs/kythe-kzip.txt [hunk 2]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+Let `v` be a `kythe.proto.VName`. Then a digest of `v` is
+
+    `v.signature NULL v.corpus NULL v.root NULL v.path NULL v.language NULL`
+
+Let `cu` be an instance of `kythe.proto.CompilationUnit`. Then a
+digest of `cu` is computed as:
+
+    `"CU" NL cu.vname NULL`
+
+followed by, for each required input:
```
Comment (justbuchanan) [inline kythe/docs/kythe-kzip.txt]: Is there a specified sort order for required inputs?

File: kythe/docs/kythe-kzip.txt [hunk 3]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+    `"RI" NL cu.required_input[i].vname NULL`
+    `"IN" NL cu.required_input[i].info.path NULL cu.required_input[i].info.digest NULL`
+
+followed by:
+
+    `"ARG" NL cu.argument[0] NULL cu.argument[1] NULL ...`
+    `"OUT" NL cu.output_key NULL`
+    `"SRC" NL cu.source_file[0] NULL cu.source_file[1] NULL ...`
+    `"CWD" NL cu.working_directory NULL`
+    `"CWD" NL cu.entry_context NULL`
```
Comment (justbuchanan) [inline kythe/docs/kythe-kzip.txt]: s/CWD/CTX/

File: kythe/docs/kythe-kzip.txt (hunk 2 above)
Comment (shahms) [reply kythe/docs/kythe-kzip.txt]: required_input is sorted and deduped by (info.digest, info.path)

File: kythe/docs/kythe-kzip.txt [hunk 4]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+
+followed by, for each `cu.environment`:
+
+    `"ENV" NL cu.environment[i].name NULL cu.environment[i].value NULL`
```
Comment (shahms) [inline kythe/docs/kythe-kzip.txt]: sorted by name

File: kythe/docs/kythe-kzip.txt [hunk 5]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+
+finally followed by, for each `cu.details`:
+
+    `"DET" NL cu.details[i].type_url NULL cu.details[i].value NULL`
```
Comment (shahms) [inline kythe/docs/kythe-kzip.txt]: sorted by type url

File: kythe/docs/kythe-kzip.txt [hunk 6]
```diff
@@ -127,3 +127,47 @@ but if set must contain the canonical JSON encoding of an `Index` message.
...
+followed by, for each required input:
+
+    `"RI" NL cu.required_input[i].vname NULL`
+    `"IN" NL cu.required_input[i].info.path NULL cu.required_input[i].info.digest NULL`
+
+followed by:
+
+    `"ARG" NL cu.argument[0] NULL cu.argument[1] NULL ...`
+    `"OUT" NL cu.output_key NULL`
+    `"SRC" NL cu.source_file[0] NULL cu.source_file[1] NULL ...`
```
Comment (shahms) [inline kythe/docs/kythe-kzip.txt]: sorted by value

**Comment by jaysachs** [comment]
The required_input de-duping seems over-specified. In particular, we could just de-dupe based on path alone (a CU with two+ required inputs for the same path but with different digests would be malformed). A case can be made that the digest algorithm can assume well-formedness, and simply sort/de-dupe based on `required_input` path.

**Comment by shahms** [comment]
> The required_input de-duping seems over-specified. In particular, we could just de-dupe based on path alone (a CU with two+ required inputs for the same path but with different digests would be malformed). A case can be made that the digest algorithm can assume well-formedness, and simply sort/de-dupe based on `required_input` path.

That seems reasonable to me.

**Pull request merged**